bundles/statistics/statsgrid/view/Diagram/Diagram.jsx

```diff
@@ -0,0 +1,357 @@
+            if (d.value < 0) {
+                return 'start';
+            }
+            return 'end';
+        })
+        .attr('x', (d) => {
+            if (d.value < 0) {
+                return 8;
+            } else {
+                return -8;
+            }
+        })
+        .attr('dy', '0.32em')
+        .style('font-size', '11px')
+        .attr('fill', '#000')
+        .text((d) => {
+            return getTextContent(d, maxLabels);
+        });
+
+    // bars
+    const bars = svg.insert('g', 'g.y').selectAll('.bar')
+        .data(chartData.data)
+        .enter()
+        .append('g')
+        .attr('class', (d) => { return d.value < 0 ? 'negative' : 'positive'; })
+        .attr('transform', (d) => {
+            return 'translate(0,' + (y(d.name) + y.bandwidth() / 2) + ')';
+        });
+
+    const barWidth = (d) => {
+        return d.value ? Math.abs(x(d.value) - x(0)) : 0;
+    };
+
+    // append rects
+    bars.append('rect')
+        .attr('class', 'bar')
+        .attr('text-anchor', 'middle')
+        .style('fill', (d, i) => { return colorScale(d.value); })
+        .attr('y', -8) // 7 is half of 15 height (pixel aligned)
+        .attr('x', (d) => { return d.value ? x(Math.min(0, d.value)) : 0; })
+        .attr('height', 17)
+        .attr('width', barWidth);
+    // append text
+    const noValStr = Oskari.getMsg('DivManazer', 'graph.noValue');
+    bars.each((d, i, nodes) => {
+        const isNumber = typeof d.value === 'number';
+        if (!valueRenderer && isNumber) {
+            return;
+        }
+        let textAnchor = 'start';
+        let transformX = '5px';
+        let locationX = x(0);
+        let rendered = noValStr;
+        let color = '#999';
+        if (isNumber) {
+            locationX = x(d.value);
+            const width = barWidth(d);
+            rendered = valueRenderer(d.value);
+            const renderedLength = typeof rendered === 'string' ? rendered.length * 8 : 0; // 8px per char (generous)
+            const fitsInBar = renderedLength < width - 10; // padding of 5px + 5px
+            if (fitsInBar) {
+                color = Oskari.util.isDarkColor(colorScale(d.value)) ? '#fff' : '#000';
+                if (d.value >= 0) {
+                    textAnchor = 'end';
+                    transformX = '-5px';
+                }
+            } else {
+                color = '#333';
+                if (d.value < 0) {
+                    textAnchor = 'end';
+                    transformX = '-5px';
+                }
+            }
+        }
+        d3.select(nodes[i])
+            .append('text')
+            .attr('x', locationX)
+            .attr('text-anchor', textAnchor)
+            .attr('dx', transformX)
+            .attr('y', 0)
+            .attr('dy', isNumber ? '0.425em' : '0.32em')
+            .style('font-size', '11px')
+            .attr('fill', color)
+            .text(rendered);
+    });
+};
+
+export const Diagram = ({ state, controller }) => {
```

Comment:
Controller isn't used? And we could break down the state at this point to just receive chartData and sortOrder as props instead of the whole state.